src/psyclone/psyir/symbols/symboltable.py

```diff
@@ -193,18 +193,48 @@ def shallow_copy(self):
         are new (symbols added to the new symbol table will not be added in
         the original but the existing objects are still the same).
 
-        :returns: a copy of this symbol table.
+        :returns: a shallow copy of this symbol table.
         :rtype: :py:class:`psyclone.psyir.symbols.SymbolTable`
 
         '''
         # pylint: disable=protected-access
         new_st = SymbolTable()
-        new_st._symbols = copy(self._symbols)
-        new_st._argument_list = copy(self._argument_list)
-        new_st._tags = copy(self._tags)
+        new_st._symbols = copy.copy(self._symbols)
+        new_st._argument_list = copy.copy(self._argument_list)
+        new_st._tags = copy.copy(self._tags)
         new_st._node = self.node
         return new_st
 
+    def deep_copy(self):
+        '''Create a copy of the symbol table where the top-level containers
+        are new and the contained symbols are also new copies. (e.g modifying
+        a symbol property will not affect the equivalent named symbol in
+        the original symbol table)
+
+        :returns: a deep copy of this symbol table.
+        :rtype: :py:class:`psyclone.psyir.symbols.SymbolTable`
+
+        '''
+        # pylint: disable=protected-access
+        new_st = SymbolTable(self.node)
+
+        # Make a copy of each symbol in the symbol table
+        for symbol in self.symbols:
+            new_st.add(symbol.copy())
+
+        # Prepare the new argument list
+        new_arguments = []
+        arg_names = [arg.name for arg in self.argument_list]
```

Comment:
Minor thing but you could do away with `arg_names` and just do it in place: `for name in [arg.name for ....]:`

Reply:
Done